### PR TITLE
713 filter group and or

### PR DESCRIPTION
**Pull request by adberglund**
## Summary 
This allows the use of `and:` and `or:` prefixes for certain filters and group_by params. To force a logical AND use the `and:` prefix as shown in the examples below. If using and, at minimum 2 values are required. This can be achieved two ways in the query params now `filter[and:account]=account1,account` or `filter[and:account]=account1&filter[and:account]=account2`

AWS:
```
GROUP_BY_OP_FIELDS = ('account', 'az', 'instance_type', 'region',
                      'service', 'storage_type', 'product_family')
FILTER_OP_FIELDS = ('account', 'service', 'region', 'az', 'product_family')
```

OCP: (uses the same fields on group by and filter)
```
OP_FIELDS = ('project', 'cluster', 'node')
```

## Examples

```
GET /local/v1/reports/openshift/compute/?filter[time_scope_value]=-1&filter[or:project]=namespace_hyper,namespace_ci
HTTP 200 OK
Allow: OPTIONS, GET
Content-Type: application/json
Vary: Accept

{
    "meta": {
        "count": 1,
        "filter": {
            "time_scope_value": "-1",
            "or:project": [
                "namespace_hyper",
                "namespace_ci"
            ]
        },
        "total": {
            "usage": {
                "value": 74.0,
                "units": "Core-Hours"
            },
            "request": {
                "value": 370.0,
                "units": "Core-Hours"
            },
            "limit": {
                "value": 370.0,
                "units": "Core-Hours"
            },
            "capacity": {
                "value": 296.0,
                "units": "Core-Hours"
            },
            "infrastructure_cost": {
                "value": 0.0,
                "units": "USD"
            },
            "derived_cost": {
                "value": 0.0,
                "units": "USD"
            },
            "cost": {
                "value": 0.0,
                "units": "USD"
            }
        }
    },
    "links": {
        "first": "/local/v1/reports/openshift/compute/?filter%5Bor%3Aproject%5D=namespace_hyper%2Cnamespace_ci&filter%5Btime_scope_value%5D=-1&limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/local/v1/reports/openshift/compute/?filter%5Bor%3Aproject%5D=namespace_hyper%2Cnamespace_ci&filter%5Btime_scope_value%5D=-1&limit=100&offset=0"
    },
    "data": [
        {
            "date": "2019-04",
            "values": [
                {
                    "date": "2019-04",
                    "usage": {
                        "value": 74.0,
                        "units": "Core-Hours"
                    },
                    "request": {
                        "value": 370.0,
                        "units": "Core-Hours"
                    },
                    "limit": {
                        "value": 370.0,
                        "units": "Core-Hours"
                    },
                    "capacity": {
                        "value": 296.0,
                        "units": "Core-Hours"
                    },
                    "infrastructure_cost": {
                        "value": 0.0,
                        "units": "USD"
                    },
                    "derived_cost": {
                        "value": 0.0,
                        "units": "USD"
                    },
                    "cost": {
                        "value": 0.0,
                        "units": "USD"
                    }
                }
            ]
        }
    ]
}
```

and produces the search filter 

```
(AND: ('usage_start__date__gte', datetime.datetime(2019, 4, 1, 0, 0, tzinfo=<UTC>)), ('usage_end__date__lte', datetime.datetime(2019, 4, 3, 0, 0, tzinfo=<UTC>)), (OR: ('namespace__icontains', 'namespace_ci'), ('namespace__icontains', 'namespace_hyper')))
```

```
GET /local/v1/reports/openshift/compute/?filter[time_scope_value]=-1&filter[and:project]=namespace_hyper,namespace_ci
HTTP 200 OK
Allow: OPTIONS, GET
Content-Type: application/json
Vary: Accept

{
    "meta": {
        "count": 1,
        "filter": {
            "time_scope_value": "-1",
            "and:project": [
                "namespace_hyper",
                "namespace_ci"
            ]
        },
        "total": {
            "usage": {
                "value": 0,
                "units": "Core-Hours"
            },
            "request": {
                "value": 0,
                "units": "Core-Hours"
            },
            "limit": {
                "value": 0,
                "units": "Core-Hours"
            },
            "capacity": {
                "value": 0.0,
                "units": "Core-Hours"
            },
            "infrastructure_cost": {
                "value": 0,
                "units": "USD"
            },
            "derived_cost": {
                "value": 0,
                "units": "USD"
            },
            "cost": {
                "value": 0,
                "units": "USD"
            }
        }
    },
    "links": {
        "first": "/local/v1/reports/openshift/compute/?filter%5Band%3Aproject%5D=namespace_hyper%2Cnamespace_ci&filter%5Btime_scope_value%5D=-1&limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/local/v1/reports/openshift/compute/?filter%5Band%3Aproject%5D=namespace_hyper%2Cnamespace_ci&filter%5Btime_scope_value%5D=-1&limit=100&offset=0"
    },
    "data": [
        {
            "date": "2019-04",
            "values": []
        }
    ]
}
```

and produces the search filter 

```
(AND: ('usage_start__date__gte', datetime.datetime(2019, 4, 1, 0, 0, tzinfo=<UTC>)), ('usage_end__date__lte', datetime.datetime(2019, 4, 3, 0, 0, tzinfo=<UTC>)), ('namespace__icontains', 'namespace_ci'), ('namespace__icontains', 'namespace_hyper'))
```

```
GET /local/v1/reports/openshift/compute/?filter[time_scope_value]=-1&group_by[or:project]=namespace_hyper,namespace_ci
HTTP 200 OK
Allow: OPTIONS, GET
Content-Type: application/json
Vary: Accept

{
    "meta": {
        "count": 1,
        "group_by": {
            "or:project": [
                "namespace_hyper",
                "namespace_ci"
            ]
        },
        "filter": {
            "time_scope_value": "-1"
        },
        "total": {
            "usage": {
                "value": 74.0,
                "units": "Core-Hours"
            },
            "request": {
                "value": 370.0,
                "units": "Core-Hours"
            },
            "limit": {
                "value": 370.0,
                "units": "Core-Hours"
            },
            "capacity": {
                "value": 296.0,
                "units": "Core-Hours"
            },
            "infrastructure_cost": {
                "value": 0.0,
                "units": "USD"
            },
            "derived_cost": {
                "value": 0.0,
                "units": "USD"
            },
            "cost": {
                "value": 0.0,
                "units": "USD"
            }
        }
    },
    "links": {
        "first": "/local/v1/reports/openshift/compute/?filter%5Btime_scope_value%5D=-1&group_by%5Bor%3Aproject%5D=namespace_hyper%2Cnamespace_ci&limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/local/v1/reports/openshift/compute/?filter%5Btime_scope_value%5D=-1&group_by%5Bor%3Aproject%5D=namespace_hyper%2Cnamespace_ci&limit=100&offset=0"
    },
    "data": [
        {
            "date": "2019-04",
            "projects": [
                {
                    "project": "namespace_ci",
                    "values": [
                        {
                            "cluster_alias": "OCP Test Provider",
                            "date": "2019-04",
                            "cluster": "nise",
                            "project": "namespace_ci",
                            "usage": {
                                "value": 37.0,
                                "units": "Core-Hours"
                            },
                            "request": {
                                "value": 185.0,
                                "units": "Core-Hours"
                            },
                            "limit": {
                                "value": 185.0,
                                "units": "Core-Hours"
                            },
                            "capacity": {
                                "value": 296.0,
                                "units": "Core-Hours"
                            },
                            "infrastructure_cost": {
                                "value": 0.0,
                                "units": "USD"
                            },
                            "derived_cost": {
                                "value": 0.0,
                                "units": "USD"
                            },
                            "cost": {
                                "value": 0.0,
                                "units": "USD"
                            }
                        }
                    ]
                },
                {
                    "project": "namespace_hyper",
                    "values": [
                        {
                            "cluster_alias": "OCP Test Provider",
                            "date": "2019-04",
                            "cluster": "nise",
                            "project": "namespace_hyper",
                            "usage": {
                                "value": 37.0,
                                "units": "Core-Hours"
                            },
                            "request": {
                                "value": 185.0,
                                "units": "Core-Hours"
                            },
                            "limit": {
                                "value": 185.0,
                                "units": "Core-Hours"
                            },
                            "capacity": {
                                "value": 296.0,
                                "units": "Core-Hours"
                            },
                            "infrastructure_cost": {
                                "value": 0.0,
                                "units": "USD"
                            },
                            "derived_cost": {
                                "value": 0.0,
                                "units": "USD"
                            },
                            "cost": {
                                "value": 0.0,
                                "units": "USD"
                            }
                        }
                    ]
                }
            ]
        }
    ]
}
```

and produces the search filter

```
(AND: ('usage_start__date__gte', datetime.datetime(2019, 4, 1, 0, 0, tzinfo=<UTC>)), ('usage_end__date__lte', datetime.datetime(2019, 4, 3, 0, 0, tzinfo=<UTC>)), (OR: ('namespace__icontains', 'namespace_ci'), ('namespace__icontains', 'namespace_hyper')))
```

```
GET /local/v1/reports/openshift/compute/?filter[time_scope_value]=-1&group_by[and:project]=namespace_hyper,namespace_ci
HTTP 200 OK
Allow: OPTIONS, GET
Content-Type: application/json
Vary: Accept

{
    "meta": {
        "count": 1,
        "group_by": {
            "and:project": [
                "namespace_hyper",
                "namespace_ci"
            ]
        },
        "filter": {
            "time_scope_value": "-1"
        },
        "total": {
            "usage": {
                "value": 0,
                "units": "Core-Hours"
            },
            "request": {
                "value": 0,
                "units": "Core-Hours"
            },
            "limit": {
                "value": 0,
                "units": "Core-Hours"
            },
            "capacity": {
                "value": 0.0,
                "units": "Core-Hours"
            },
            "infrastructure_cost": {
                "value": 0,
                "units": "USD"
            },
            "derived_cost": {
                "value": 0,
                "units": "USD"
            },
            "cost": {
                "value": 0,
                "units": "USD"
            }
        }
    },
    "links": {
        "first": "/local/v1/reports/openshift/compute/?filter%5Btime_scope_value%5D=-1&group_by%5Band%3Aproject%5D=namespace_hyper%2Cnamespace_ci&limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/local/v1/reports/openshift/compute/?filter%5Btime_scope_value%5D=-1&group_by%5Band%3Aproject%5D=namespace_hyper%2Cnamespace_ci&limit=100&offset=0"
    },
    "data": [
        {
            "date": "2019-04",
            "projects": []
        }
    ]
}
```

and produces the search filter 

```
(AND: ('usage_start__date__gte', datetime.datetime(2019, 4, 1, 0, 0, tzinfo=<UTC>)), ('usage_end__date__lte', datetime.datetime(2019, 4, 3, 0, 0, tzinfo=<UTC>)), ('namespace__icontains', 'namespace_ci'), ('namespace__icontains', 'namespace_hyper'))
```

### CODE REVIEW

File: koku/api/report/queries.py [hunk 1]
```diff
@@ -993,11 +993,14 @@ def _get_search_filter(self, filters):
...
 
+        # Update filters that specifiy and or or in the query parameter
+        filters = self._set_operator_specified_filters(filters, 'and')
+        filters = self._set_operator_specified_filters(filters, 'or')
         # Update filters with tag filters
         filters = self._set_tag_filters(filters)
 
         composed_filters = filters.compose()
-        LOG.debug(f'_get_search_filter: {composed_filters}')
+        LOG.info(f'_get_search_filter: {composed_filters}')
```
Comment (adberglund) [inline koku/api/report/queries.py]:
```suggestion
        LOG.debug(f'_get_search_filter: {composed_filters}')
```

File: koku/api/report/queries.py [hunk 2]
```diff
@@ -1031,6 +1034,36 @@ def _set_tag_filters(self, filters):
                 filters.add(q_filter)
         return filters
 
+    def _set_operator_specified_filters(self, filters, operator):
```
Comment (dccurtis) [inline koku/api/report/queries.py]: Can this move to `api.query_filter.QueryFilter` to de-duplicate it from the tag query handler?

File: koku/api/report/aws/serializers.py [hunk 3]
```diff
@@ -71,6 +60,15 @@ def __init__(self, *args, **kwargs):
             # Add AWS tag keys to allowable fields
             self.fields.update(tag_keys)
 
+        and_fields = {'and:' + field: StringOrListField(child=serializers.CharField(),
```
Comment (dccurtis) [inline koku/api/report/aws/serializers.py]: This block for and_fields, or_fields, self.fields.update() seem to be the same for all `GroupbySerializer` and `FilterSerializer`  Would making an `api.serializers.py` to create a base class consolidate this?

Comment (blentz) [reply koku/api/report/aws/serializers.py]: Agree with Doug. A common super-class probably makes sense here.

**Comment by adberglund** [comment]
@dccurtis @blentz I went with a function call so that all those and/or dictionary updates are at least consolidated. Common base class didn't quite work as OCP on AWS inherits from AWS and then adds the extra OCP fields on top of that. And this keeps us from requiring an argument on the filter and group by serializer class to instantiate which is more in line with the `def __init__(self, *args, **kwargs):` style that DRF has.

**Pull request merged**